docs/adr/adr-11-blocksync-overhaul.md

```diff
@@ -0,0 +1,329 @@
+for managing any CAR files with indexing, convenient abstractions, tools, recovery mechanisms, etc.
+
+- __LNs DASing remains untouched__. Both the networking protocol and storage for LNs remains untouched as it fulfills
+    the requirements. This includes Bitswap as backbone protocol for requesting samples and global Badger KVStore.
+
+- __New libp2p based Exchange protocol is introduced for data/share exchange purposes.__ FNs/BNs servers LNs clients.
+
+### Detailed Design
+
+> All the comments on the API definitions should be preserved and potentially improved by implementations.
+
+#### Block/EDS Storage
+
+The new block storage design is solely additive. Meaning that all the existing storage related components and functionality
+are kept with additional components introduced. Altogether, existing and new components will be recomposed to serve the
+foundation of our improved block storage subsystem.
+
+The central data structure representing Celestia block data is EDS(`rsmt2d.ExtendedDataSquare`) and the new storage design
+is focused around storing entire EDSes as a whole rather than a set of individual chunks s.t. storage subsystem
+can handle storing and streaming/serving blocks of 4mb sizes and more.
+
+##### EDS Serde
+
+Storing EDS as a whole requires EDS (de)serialization. For this the [CAR format](https://ipld.io/specs/transport/car) is
+chosen.
+
+###### `share.WriteEDS`
+
+To write EDS into a stream/file `WriteEDS` is introduced.
+// TODO Describe logic
+
+NOTE: CAR provides [a utility](https://github.com/ipld/go-car/blob/master/car.go#L47) to serialize any DAG into the file and
+there is a way to serialize EDS into DAG(`share/ipld.ImportShares`). This approach is the simplest and traverses shares
+and Merkle Proofs in depth-first manner packing them in a CAR file. However, this is incompatible with the requirement
+to truncate the CAR file to get strictly the first quadrant out of it without NMT proofs, so serialization must be
+different from the utility to support that.
+
+NOTE2: Alternatively to `WriteEDS`, and `EDSReader` could be introduced to make EDS-to-stream handling more idiomatic
+and efficient in some cases, with the cost of more complex implementation.
+
+```go
+// WriteEDS writes whole EDS into given io.Writer as CARv1 file.
+// All its shares and recomputed NMT proofs.
+func WriteEDS(context.Context, *rsmt2d.ExtendedDataSquare, io.Writer) error
+```
+
+###### `share.ReadEDS`
+
+To read EDS out of stream/file `ReadEDS` is introduced. Internally, it
+
+- Imports EDS with an empty pre-allocated slice. NOTE: Size can be taken from DataRoot
+- Wraps given io.Reader with [`BlockReader`](https://github.com/ipld/go-car/blob/master/v2/block_reader.go#L16)
+- Reads out blocks one by one and fills up the EDS quadrant via `EDS.SetCell`
+- Recomputes and validates via `EDS.Repair`
+
+```go
+// ReadEDS reads EDS quadrant(1/4) from io.Reader CAR file.
+//
+// It expects strictly first EDS quadrant(top left).
```

Comment:
I do not think that ReadEDS should be passed an io.Reader that is already responsible for only reading out the part of the file that ReadEDS needs - ReadEDS sounds like it should handle this logic itself

Reply:
Ok I read further down that `GetCar` should be responsible for this - I still think GetCAR should return a reader over the entire CAR then

Reply:
This Reader is the data that is served over the wire. If we make Reader read the whole EDS, then it means the whole EDS is transferred over the wire, which is what we want to avoid. We can, though, engineer something smart that Reads only quadrant out of full Reader, but that's tricky, adds complexity(not justified), and anyway, this logic has to live somewhere, like in Store, or in protocol

Reply:
yeah that complexity would not be justified true

just hearing "This Reader is the data that is served over the wire." changed my mind on this

Reply:
Resolved then

Reply:
Additionally added a note on this